src/rdhttp.c

```diff
@@ -223,6 +223,155 @@ rd_http_error_t *rd_http_get (const char *url, rd_buf_t **rbufp) {
+ *
+ * @returns The token will be responsed in \p *res.
+ *          Returns NULL on success or an error
+ *          object on error - this error object must be destroyed
+ *          by calling rd_http_error_destroy().
+ */
+rd_http_error_t *rd_http_extract_jwt_from_json (cJSON **cjson,
+                                               char **res,
+                                               const char *key) {
+        char *token;
+        cJSON *jval;
+        cJSON *parsed_token = NULL;
+        rd_http_error_t *herr;
+
+        cJSON_ArrayForEach(jval, *cjson) {
+                parsed_token = cJSON_GetObjectItem(*cjson, key);
+                break;
+        }
+        if (parsed_token == NULL) {
+                herr = rd_http_error_new(-1,
+                        "Expected non-empty JSON response");
+                return herr;
+        }
+
+        token = cJSON_Print(parsed_token);
+        *res = rd_malloc(strlen(token) + 1);
+        strcpy(*res, token);
+        rd_free(token);
+        return NULL;
+}
+
+
+/**
+ * @brief Extract the json string that includes jwt from \p hreq.
+ *
+ * @returns Returns the response (even if there's a HTTP error code returned)
+ *          in \p *json.
+ */
+rd_http_error_t *rd_http_parse_token_to_json (rd_http_req_t *hreq,
+                                              cJSON **jsonp) {
+        size_t len = 0;
+        char *raw_json = NULL;
+        const char *end = NULL;
+        rd_slice_t slice;
+        rd_http_error_t *herr = NULL;
+        *jsonp = NULL;
+
+        /* cJSON requires the entire input to parse in contiguous memory. */
+        rd_slice_init_full(&slice, hreq->hreq_buf);
+        len = rd_buf_len(hreq->hreq_buf);
+
+        raw_json = rd_malloc(len + 1);
+        rd_slice_read(&slice, raw_json, len);
+        raw_json[len] = '\0';
+
+        /* Parse JSON */
+        end = NULL;
+        *jsonp = cJSON_ParseWithOpts(raw_json, &end, 0);
+
+        if (!*jsonp && !herr)
+                herr = rd_http_error_new(hreq->hreq_code,
+                                         "Failed to parse JSON response "
+                                         "at %"PRIusz"/%"PRIusz,
+                                         (size_t)(end - raw_json), len);
+        rd_free(raw_json);
+        rd_http_req_destroy(hreq);
+        return herr;
+}
+
+
+/**
+ * @brief Perform a blocking HTTP(S) request to \p url with HTTP(S)
+ *        headers and data with 20s timeout.
+ *        If the HTTP(S) request fails, will retry another \p retry times.
+ *
+ * @returns The result will be responsed in \p *jsonp.
+ *          Returns NULL on success (HTTP response code < 400), or an error
+ *          object on transport or HTTP error - this error object must be
+ *          destroyed by calling rd_http_error_destroy().
+ *
+ * @locality Any thread.
+ */
+rd_http_error_t
+*rd_http_post_json (const char *url,
+                    struct curl_slist **headers,
+                    const char *data_to_token,
```

Comment:
Hi @edenhill, thanks for reviewing.

Yeah, the function is actually post headers and data and expect json response, sorry for the naming.

For the content-type, it's one of fields in the header, the default one from the python HTTP server is "application/x-www-form-urlencoded", same as our expectation. So I didn't add it to the header, but I will add it in case no surprise from other provider.

Reply:
Sounds good, we want to be explicit and not rely on one server's default value :+1: